synapse/lib/stormlib/stix.py

```diff
@@ -0,0 +1,818 @@
+
+        s_stormtypes.Prim.__init__(self, 'bundle', path=path)
+
+        self.runt = runt
+        self.libstix = libstix
+        self.config = config
+        self.locls.update(self.getObjLocals())
+        self.objs = {}  # id -> STIX obj(dict)
+        self.maxsize = config.get('maxsize', 10000)
+
+    async def value(self):
+        return self.pack()
+
+    def getObjLocals(self):
+        return {
+            'add': self.add,
+            'pack': self.pack,
+        }
+
+    # TODO config modification helpers
+    #async def addPropMap(self, formname, stixtype, propname, stormtext):
+    #async def addRelsMap(self, formname, stixtype, relname, targtype,  stormtext):
+
+    async def add(self, node, stixtype=None):
+
+        if len(self.objs) >= self.maxsize:
+            mesg = f'STIX Bundle is at maxsize ({self.maxsize}).'
+            raise s_exc.StormRuntimeError(mesg=mesg)
+
+        if not isinstance(node, s_node.Node):
+            await self.runt.warnonce('STIX bundle add() method requires a node.')
+            return None
+
+        formconf = self.config['forms'].get(node.form.name)
+        if formconf is None:
+            await self.runt.warnonce('STIX bundle has no config for mapping {node.form.name}.')
+            return None
+
+        if stixtype is None:
+            stixtype = formconf.get('default')
+
+        # cyber observables have UUIDv5 the rest have UUIDv4
+        if stixtype in stix_observables:
+            stixid = f'{stixtype}--{uuid5(node.ndef)}'
+        else:
+            stixid = f'{stixtype}--{uuid4(node.ndef)}'
+
+        if self.objs.get(stixid) is not None:
+            return stixid
+
+        stixconf = formconf['stix'].get(stixtype)
+        if stixconf is None:
+            await self.runt.warnonce(f'STIX bundle config has no config to map {node.form.name} to {stixtype}.')
+            return None
+
+        stixitem = self.objs.get(stixid)
+        if stixitem is None:
+            stixitem = self.objs[stixid] = self._initStixItem(stixid, stixtype, node)
+
+        props = stixconf.get('props')
+        if props is not None:
+
+            for name, storm in props.items():
+                valu = await self._callStorm(storm, node)
+                if valu is s_common.novalu:
+                    continue
+
+                stixitem[name] = valu
+
+        for (relname, reltype, relstorm) in stixconf.get('rels', ()):
+            async for relnode, relpath in node.storm(self.runt, relstorm):
+                n2id = await self.add(relnode, stixtype=reltype)
+                await self._addRel(stixid, relname, n2id)
+
+        return stixid
+
+    def _initStixItem(self, stixid, stixtype, node):
+        retn = {
+            'id': stixid,
+            'type': stixtype,
+            'spec_version': '2.1',
+        }
+        return retn
+
+    async def _addRel(self, srcid, reltype, targid):
+
+        stixid = f'relationship--{uuid4((srcid, targid))}'
```

Comment:
It isn't explicit what the cardinality of a particular relationship is, but it makes sense that you can have more than more relationship type between two stix nodes (like threat actor -> identity). I'd suggest adding reltype to the uuid4.

Reply:
great catch.  thanks!